games/xworld3d/x3item.h

```diff
@@ -0,0 +1,197 @@
+
+    virtual ~X3Item() {
+        this->destroy();
+    }
+
+    const Object& object() {
+        return object_;
+    }
+
+    Pose pose() { return object_.pose(); }
+
+    void set_pose(const Pose& pose);
+
+    void set_speed(double vx, double vy, double vz);
+
+    void set_pose_and_speed(const Pose& pose, double vx, double vy, double vz);
+
+    bool equal(const X3Item* i) {
+        return (name_ == i->name());
+    }
+
+    void destroy() override;
+
+protected:
+    std::string model_file_;
+    Object object_;
+};
+
+/*********************************** X3Block **********************************/
+class X3Block : public X3Item {
+public:
+    X3Block(const X3ItemInfo& info, World& world) : X3Item(info, world) {
+        CHECK(info.type == X3EntityType::BLOCK);
+    }
+};
+
+typedef std::shared_ptr<X3Block> X3BlockPtr;
+
+/*********************************** X3Goal ***********************************/
+class X3Goal : public X3Item {
+public:
+    X3Goal(const X3ItemInfo& info, World& world) : X3Item(info, world) {
+        CHECK(info.type == X3EntityType::GOAL);
+    }
+};
+
+typedef std::shared_ptr<X3Goal> X3GoalPtr;
+
+/*********************************** X3Agent **********************************/
+class X3Agent : public X3Item {
+public:
+    X3Agent(const X3ItemInfo& info, World& world,
+            float speed_norm, int orientation_bins,
+            float reaching_dist);
+
+    void move_forward();
+
+    void turn_left();
+
+    void turn_right();
+
+    void jump();
+
+    void get_direction(double &dir_x, double &dir_y) {
+        dir_x = dir_x_;
+        dir_y = dir_y_;
+    }
+
+    float reach_test(const Pose& gpose);
+
+private:
+    void set_direction();
+
+    const float speed_norm_;
+    int yaw_id_;
+    double dir_x_;
+    double dir_y_;
+    const int orientation_bins_;
+    const float reaching_dist_; // An agent can collect this goal if it is
+                                // within the reaching distance of this goal
+};
+
+typedef std::shared_ptr<X3Agent> X3AgentPtr;
+
+/********************************** X3Camera **********************************/
+class X3Camera {
+public:
+    X3Camera(World& world, int img_height, int img_width);
+
+    Pose pose() { return camera_.pose(); }
+    
+    // Return the image seen by agent. 
+    // If bird_view is true, a bird view image is also returned.
+    roboschool::RenderResult render(X3Agent* agent, bool bird_view = false);
+
+    // Camera can be attached to an agent so that the rendered image is centered
+    // at the agent
+    void attach_agent(X3Agent* agent);
+
+    void detach() { agent_ = NULL; }
+
+private:
+    // Update the pose of the camera.
+    void update(bool bird_view);
+
+    Camera camera_;
+    X3Agent* agent_;
+};
+
+}} // simulator::xworld3d
+
+// TODO: need a better hash function
+namespace std {
+template <>
+struct hash<simulator::Vec3> {
+    size_t operator()(const simulator::Vec3& l) const {
```

Comment:
make sure you fix this before checking in. This error will be a disaster if left for the future.